core-bundle/src/Resources/contao/modules/Module.php

```diff
@@ -351,89 +351,104 @@ protected function renderNavigation($pid, $level=1, $host=null, $language=null)
 						break;
 				}
 
-				$row = $objSubpage->row();
-				$trail = \in_array($objSubpage->id, $objPage->trail);
+				$items[] = $this->compileNavigationRow($objPage, $objSubpage, $subitems, $href);
+			}
+		}
 
-				// Use the path without query string to check for active pages (see #480)
-				list($path) = explode('?', Environment::get('request'), 2);
+		// Add classes first and last
+		if (!empty($items))
+		{
+			$last = \count($items) - 1;
 
-				// Active page
-				if (($objPage->id == $objSubpage->id || ($objSubpage->type == 'forward' && $objPage->id == $objSubpage->jumpTo)) && !($this instanceof ModuleSitemap) && $href == $path)
-				{
-					// Mark active forward pages (see #4822)
-					$strClass = (($objSubpage->type == 'forward' && $objPage->id == $objSubpage->jumpTo) ? 'forward' . ($trail ? ' trail' : '') : 'active') . (($subitems != '') ? ' submenu' : '') . ($objSubpage->protected ? ' protected' : '') . ($objSubpage->cssClass ? ' ' . $objSubpage->cssClass : '');
+			$items[0]['class'] = trim($items[0]['class'] . ' first');
+			$items[$last]['class'] = trim($items[$last]['class'] . ' last');
+		}
 
-					$row['isActive'] = true;
-					$row['isTrail'] = false;
-				}
+		$objTemplate->items = $items;
 
-				// Regular page
-				else
-				{
-					$strClass = (($subitems != '') ? 'submenu' : '') . ($objSubpage->protected ? ' protected' : '') . ($trail ? ' trail' : '') . ($objSubpage->cssClass ? ' ' . $objSubpage->cssClass : '');
+		return !empty($items) ? $objTemplate->parse() : '';
+	}
 
-					// Mark pages on the same level (see #2419)
-					if ($objSubpage->pid == $objPage->pid)
-					{
-						$strClass .= ' sibling';
-					}
+	/**
+	 * Compile the navigation row and return it as array
+	 *
+	 * @param PageModel $objPage
+	 * @param PageModel $objSubpage
+	 * @param string    $subitems
+	 * @param string    $href
+	 *
+	 * @return array
+	 */
+	protected function compileNavigationRow($objPage, $objSubpage, $subitems, $href)
```

Comment:
```suggestion
	protected function compileNavigationRow(PageModel $objPage, PageModel $objSubpage, $subitems, $href)
```

Reply:
When creating the PR I was not used to the fact that TypeHints can also be used in legacy code ;-)
Should we also add the `string` hint for the last parameters, shouldn't we?

Reply:
No, we only use type hints that were already supported in PHP 5.6 in the legacy code.